### PR TITLE
LOOP-1406: Add authentication when saving therapy settings

### PR DESCRIPTION
**Pull request by rickpasetto**
This refactors a bit of the "saving" logic to have common code for presenting the authentication challenge, and then presenting an alert if the challenge fails.

https://tidepool.atlassian.net/browse/LOOP-1406

### CODE REVIEW

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (mpangburn) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: [nit] I don't think this typealias is necessary

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (mpangburn) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: [nit] Typically an instance of Void is spelled `()` (the empty tuple)

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (mpangburn) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: Should have an `else` branch that calls `completion` successfully if the policy can't be evaluated

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (mpangburn) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: [nit] Perhaps worth asserting that `error` is non-nil if `success` is `false`

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (mpangburn) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: I think this can be `fileprivate`

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (mpangburn) [inline LoopKitUI/Views/ScheduleEditor.swift]: This suggestion might sound odd, but I'm actually not sure we should present an alert when authentication fails here. Since an incorrect biometry reading keeps the user in the authentication context; generally they proceed only by 1) correctly authenticating or 2) cancelling. We can defer communication of auth errors to the auth modal itself.

File: LoopKitUI/Views/Settings Editors/PresentedAlert.swift
Comment (mpangburn) [inline LoopKitUI/Views/Settings Editors/PresentedAlert.swift]: If we defer auth errors to the modal, we can return this to the ScheduleEditor namespace. If you want to extract it, I think we should rename it for clarity; something like SettingSaveAlert.

**Review by mpangburn — COMMENTED** [review]
The general approach is nice! I believe we can simplify by deferring auth error presentation to the iOS modal.

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (rickpasetto) [reply LoopKitUI/Extensions/Environment+Authenticate.swift]: Ok.  I think someday we may want it ;)

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (rickpasetto) [reply LoopKitUI/Extensions/Environment+Authenticate.swift]: ooh good catch!

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (rickpasetto) [reply LoopKitUI/Views/ScheduleEditor.swift]: I don't know... I think this is a design call.  Normally I'd agree, but we're being ultra-conservative when it comes to communicating to the user what happened.  For example, if the user cancels the authentication, this will pop up an error:
Simulator Screen Shot - iPhone SE (2nd generation) - 2020-08-03 at 15 07 30

Pedantic?  Yes...but maybe not a bad idea?  I suggest we try leaving it as-is here and see how design review goes.  WDYT?

File: LoopKitUI/Views/Settings Editors/PresentedAlert.swift
Comment (rickpasetto) [reply LoopKitUI/Views/Settings Editors/PresentedAlert.swift]: I'll rename it... but I kind of like it as a reusable enum...

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (novalegra) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: [nit/q] If we're asserting the error is `nil` here, then would it make sense to force unwrap it on the next line?

**Review by novalegra — APPROVED** [review]
One q, otherwise LGTM!

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (mpangburn) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: Ah, just noticed this extension is public. I'm not sure it makes sense for LoopKitUI to vend this as public API.

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (mpangburn) [reply LoopKitUI/Extensions/Environment+Authenticate.swift]: I could see it either way; the assert won't be triggered in release builds, so if the framework had a bug where the assertion failed having a fallback isn't unreasonable.

Comment (mpangburn) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: I'm not sure failure is the correct outcome if the policy can't be evaluated. What if the error is `biometryNotAvailable` or `biometryNotEnrolled`? If a user doesn't have auth set up, they should still be able to save their settings.

The precedent here (in bolusing) is to proceed as successful if the policy can't be evaluated.

Comment (rickpasetto) [reply LoopKitUI/Extensions/Environment+Authenticate.swift]: `assert` is compiled out in production code (https://developer.apple.com/documentation/swift/1541112-assert)
(note: `precondition` is not)
force-unwrapping would cause a crash in production code.  I felt it was unnecessary to do that, though it would not be unreasonable given that it would indicate Apple's APIs would violate their own contract, which is unlikely.  I could go either way.

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (mpangburn) [inline LoopKitUI/Views/ScheduleEditor.swift]: I'll continue to push back on presenting an error here. All error states are communicated by the auth modal. If the user pushes cancel, they should not see a modal that tells them they just cancelled. I weighed in on the Slack thread too; seems that Pete agrees we don't need to handle the presentation of an error here.

**Review by mpangburn — COMMENTED** [review]
If auth is unavailable, I believe the outcome should be `success`.

I continue to think we don't need to handle presentation of auth errors, but won't block on that.

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (rickpasetto) [reply LoopKitUI/Views/ScheduleEditor.swift]: Okay, I will remove it.  That said, would you be okay with my leaving this new `SettingSaveAlert` in place as a generally-usable facility?

Comment (mpangburn) [reply LoopKitUI/Views/ScheduleEditor.swift]: Yeah, I don't see a problem with that

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (rickpasetto) [reply LoopKitUI/Extensions/Environment+Authenticate.swift]: Good point.  I'll ask Tejasvi in Slack what the behavior should be.

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (rickpasetto) [reply LoopKitUI/Views/ScheduleEditor.swift]: ;) I got rid of it.  Sorry...Please take another look, @mpangburn ?

File: LoopKitUI/Views/Settings Editors/DeliveryLimitsEditor.swift
Comment (mpangburn) [inline LoopKitUI/Views/Settings Editors/DeliveryLimitsEditor.swift]: `print` can be deleted

File: LoopKitUI/Views/Settings Editors/SuspendThresholdEditor.swift
Comment (mpangburn) [inline LoopKitUI/Views/Settings Editors/SuspendThresholdEditor.swift]: `print` can be deleted

File: LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift
Comment (mpangburn) [inline LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift]: `print` can be deleted

File: LoopKitUI/Views/Settings Editors/CorrectionRangeOverridesEditor.swift
Comment (mpangburn) [inline LoopKitUI/Views/Settings Editors/CorrectionRangeOverridesEditor.swift]: `print` can be deleted

**Review by mpangburn — APPROVED** [review]
You can also remove the 'cancel' and 'ok' properties that were added to AlertContent now, I believe

**Comment by rickpasetto** [comment]
Sorry to do this to you @mpangburn and @novalegra , but I received feedback on a slack thread that we need to change the "back" button behavior of the editors to show `Cancel` instead, so that it is clear users that their settings change will be discarded.  So, if you could please look at the new changes above, I'd really appreciate it. :)

**Comment by rickpasetto** [comment]
@mpangburn @novalegra ...and, another update.  Double apologies again, but I now have learned that the expected behavior is for the "Cancel" button to only appear when the value has changed, otherwise it _should_ show the Back (previous screen) button.  So, I've updated it...please have another look.

File: LoopKitUI/Extensions/Environment+Authenticate.swift
Comment (novalegra) [inline LoopKitUI/Extensions/Environment+Authenticate.swift]: [nit] Missing a space

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (novalegra) [inline LoopKitUI/Views/ScheduleEditor.swift]: I don't think we want the cancel button to be in the acceptance flow - users should be able to go back and forth between screens (I discussed the "close" button issue with Melissa and Brooke, and we all agreed that the editors are too crowded to have one)

File: LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift
Comment (novalegra) [inline LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift]: There's no title for the other configurations of this screen, which makes it so no title appears in the acceptance flow or the new therapy settings screen.

**Review by novalegra — CHANGES_REQUESTED** [review]
There are some changes that affect titles/the acceptance flow that we should probably fix before this lands.

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (rickpasetto) [reply LoopKitUI/Views/ScheduleEditor.swift]: Good catch.  I'll fix it! 👍

**Review by mpangburn — COMMENTED** [review]
I think the presence of `cancel` being plumbed all the way through is a sign that it should go through the environment. You can pass it through using the `dismiss` key.

File: LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift
Comment (rickpasetto) [reply LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift]: I was thrown by the fact that sometimes the title "Insulin Model" got "stuck" sometimes when navigating back.  I thought this would fix it, but turns out this is wrong, too.  I can't reproduce it, so I'll put this back to where it used to be.  But I'll keep an eye out for it.

**Comment by rickpasetto** [comment]
@mpangburn good idea.  I'll see if I can do that.

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (novalegra) [inline LoopKitUI/Views/ScheduleEditor.swift:160]: [nit] spacing

**Review by novalegra — APPROVED** [review]
LGTM, thanks Rick!

File: LoopKitUI/Views/ScheduleEditor.swift
Comment (rickpasetto) [reply LoopKitUI/Views/ScheduleEditor.swift:160]: This is indented how Xcode wants it.  I could reduce the vertical space used...

Comment (novalegra) [reply LoopKitUI/Views/ScheduleEditor.swift:160]: Gotcha - no need to reduce vertical space :)

**Review by mpangburn — APPROVED** [review]
Looks great!

**Pull request merged**
